unidist/core/backends/mpi/core/communication.py

```diff
@@ -347,27 +534,119 @@ def isend_complex_operation(comm, operation_type, operation_data, dest_rank):
         Data object to send.
     dest_rank : int
         Target MPI process to transfer data.
+    is_serialized : bool
+        `operation_data` is already serialized or not.
 
     Returns
     -------
-    list
-        A list of pairs, ``MPI_Isend`` handler and associated data to send.
+    dict or None
+        Serialization data for caching purpose.
     """
-    handler_list = []
     # Send operation type
+    mpi_send_object(comm, operation_type, dest_rank)
+    # Send operation data
+    return send_operation_data(comm, operation_data, dest_rank, is_serialized)
+
+
+def isend_complex_operation(
+    comm, operation_type, operation_data, dest_rank, is_serialized=False
+):
+    """
+    Send operation and data that consist of different user provided complex types, lambdas and buffers.
+
+    Non-blocking asynchronous interface.
+    The data is serialized with ``unidist.core.backends.mpi.core.ComplexSerializer``.
+    Function works with already serialized data.
+
+    Parameters
+    ----------
+    comm : object
+        MPI communicator object.
+    operation_type : ``unidist.core.backends.mpi.core.common.Operation``
+        Operation message type.
+    operation_data : object
+        Data object to send.
+    dest_rank : int
+        Target MPI process to transfer data.
+    is_serialized : bool
+        `operation_data` is already serialized or not.
+
+    Returns
+    -------
+    dict or None
+        Serialization data for caching purpose.
+    """
+    # Send operation type
+    handler_list = []
     h1 = mpi_isend_object(comm, operation_type, dest_rank)
     handler_list.append((h1, None))
-    # Send complex dictionary data
-    h2_list = isend_complex_data(comm, operation_data, dest_rank)
-    handler_list.extend(h2_list)
-    return handler_list
+
+    # Send operation data
+    if is_serialized:
+        # Send already serialized data
+        s_data = operation_data["s_data"]
+        raw_buffers = operation_data["raw_buffers"]
+        len_buffers = operation_data["len_buffers"]
+
+        h2_list = isend_complex_data_impl(
```

Comment:
Possible, the only consumer of `isend_complex_data_impl` should be `isend_complex_data` function. It looks strange when in one branch you call `isend_complex_data_impl`, but in other `isend_complex_data`. I believe that this if-else can be moved to `isend_complex_data` implementation